File: src/main/java/at/yawk/math/data/UnaryExpression.java
package at.yawk.math.data;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * @author yawkat
 */
@RequiredArgsConstructor
@EqualsAndHashCode
@Getter
public abstract class UnaryExpression implements Expression {
    private final Expression child;

    protected abstract String toString(String content);

    @Override
    public String toString(int radix) {
        return toString(child.toString(radix));
    }

    @Override
    public String toString() {
        return toString(10);
    }
}

File: src/main/kotlin/at/yawk/math/data/UnaryExpression.kt
package at.yawk.math.data

/**
 * @author yawkat
 */
abstract class UnaryExpression(val child: Expression) : BaseExpression() {
    protected abstract fun toString(content: String): String

    override fun toString(radix: Int): String {
        return toString(child.toString(radix))
    }

    abstract override fun equals(other: Any?): Boolean
    abstract override fun hashCode(): Int
}